{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "all_kat_hist                            freq_k31.hist\n",
      "all_kat_hist.dist_analysis.json         freq_k7.hist\n",
      "\u001b[0m\u001b[01;35mall_kat_hist.png\u001b[0m                        \u001b[01;34mimages\u001b[0m\n",
      "\u001b[01;34mbigsi\u001b[0m                                   kat.hist\n",
      "\u001b[01;34mdata\u001b[0m                                    kat.hist.dist_analysis.json\n",
      "ERR1144974                              \u001b[01;35mkat.hist.png\u001b[0m\n",
      "ERR1144974.dist_analysis.json           microbial-bigsi.ipynb\n",
      "ERR1144974_kat_hist                     microbial-genomes.txt\n",
      "ERR1144974_kat_hist.dist_analysis.json  \u001b[01;35mo.png\u001b[0m\n",
      "\u001b[01;35mERR1144974_kat_hist.png\u001b[0m                 process-data.ipynb\n",
      "\u001b[01;35mERR1144974.png\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "NOTEBOOK_DIR=`git rev-parse --show-toplevel`\n",
    "ROOT_DIR=$NOTEBOOK_DIR/microbial\n",
    "cd $ROOT_DIR\n",
    "ls"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Microbial BIGSI\n",
    "\n",
    "This runs `bigsi` <https://bigsi.readme.io/> on the microbial dataset. First, let's setup some variables to define size of bloom filters/bigsi parameters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "number_hashes=3\n",
    "m_value=25000000\n",
    "kmer_size=9\n",
    "\n",
    "threads=32"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 1: Counting k-mers\n",
    "\n",
    "Step 1 involves counting kmers using [mccortex](https://github.com/mcveanlab/mccortex)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘bigsi’: File exists\n",
      "~/workspace/comp7934-project/microbial/bigsi ~/workspace/comp7934-project/microbial ~/workspace/comp7934-project/microbial ~/workspace/comp7934-project/microbial ~/workspace/comp7934-project/microbial ~/workspace/comp7934-project/microbial ~/workspace/comp7934-project/microbial ~/workspace/comp7934-project/microbial\n",
      "Working on ERR1144974\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s ERR1144974 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/ERR1144974_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/ERR1144974_2.fastq.gz ERR1144974.ctx\n",
      "+ set +x\n",
      "Working on ERR1144975\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s ERR1144975 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/ERR1144975_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/ERR1144975_2.fastq.gz ERR1144975.ctx\n",
      "+ set +x\n",
      "Working on ERR1144976\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s ERR1144976 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/ERR1144976_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/ERR1144976_2.fastq.gz ERR1144976.ctx\n",
      "+ set +x\n",
      "Working on ERR1144977\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s ERR1144977 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/ERR1144977_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/ERR1144977_2.fastq.gz ERR1144977.ctx\n",
      "+ set +x\n",
      "Working on ERR1144978\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s ERR1144978 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/ERR1144978_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/ERR1144978_2.fastq.gz ERR1144978.ctx\n",
      "+ set +x\n",
      "Working on ERR3655992\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s ERR3655992 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/ERR3655992_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/ERR3655992_2.fastq.gz ERR3655992.ctx\n",
      "+ set +x\n",
      "Working on ERR3655994\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s ERR3655994 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/ERR3655994_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/ERR3655994_2.fastq.gz ERR3655994.ctx\n",
      "+ set +x\n",
      "Working on ERR3655996\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s ERR3655996 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/ERR3655996_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/ERR3655996_2.fastq.gz ERR3655996.ctx\n",
      "+ set +x\n",
      "Working on ERR3655998\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s ERR3655998 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/ERR3655998_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/ERR3655998_2.fastq.gz ERR3655998.ctx\n",
      "+ set +x\n",
      "Working on ERR3656002\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s ERR3656002 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/ERR3656002_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/ERR3656002_2.fastq.gz ERR3656002.ctx\n",
      "+ set +x\n",
      "Working on ERR3656004\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s ERR3656004 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/ERR3656004_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/ERR3656004_2.fastq.gz ERR3656004.ctx\n",
      "+ set +x\n",
      "Working on ERR3656010\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s ERR3656010 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/ERR3656010_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/ERR3656010_2.fastq.gz ERR3656010.ctx\n",
      "+ set +x\n",
      "Working on ERR3656012\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s ERR3656012 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/ERR3656012_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/ERR3656012_2.fastq.gz ERR3656012.ctx\n",
      "+ set +x\n",
      "Working on ERR3656013\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s ERR3656013 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/ERR3656013_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/ERR3656013_2.fastq.gz ERR3656013.ctx\n",
      "+ set +x\n",
      "Working on ERR3656015\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s ERR3656015 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/ERR3656015_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/ERR3656015_2.fastq.gz ERR3656015.ctx\n",
      "+ set +x\n",
      "Working on ERR3656018\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s ERR3656018 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/ERR3656018_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/ERR3656018_2.fastq.gz ERR3656018.ctx\n",
      "+ set +x\n",
      "Working on ERR3656019\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s ERR3656019 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/ERR3656019_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/ERR3656019_2.fastq.gz ERR3656019.ctx\n",
      "+ set +x\n",
      "Working on SRR10298903\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s SRR10298903 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10298903_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10298903_2.fastq.gz SRR10298903.ctx\n",
      "+ set +x\n",
      "Working on SRR10298904\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s SRR10298904 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10298904_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10298904_2.fastq.gz SRR10298904.ctx\n",
      "+ set +x\n",
      "Working on SRR10298905\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s SRR10298905 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10298905_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10298905_2.fastq.gz SRR10298905.ctx\n",
      "+ set +x\n",
      "Working on SRR10298906\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s SRR10298906 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10298906_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10298906_2.fastq.gz SRR10298906.ctx\n",
      "+ set +x\n",
      "Working on SRR10298907\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s SRR10298907 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10298907_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10298907_2.fastq.gz SRR10298907.ctx\n",
      "+ set +x\n",
      "Working on SRR10512964\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s SRR10512964 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10512964_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10512964_2.fastq.gz SRR10512964.ctx\n",
      "+ set +x\n",
      "Working on SRR10512965\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s SRR10512965 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10512965_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10512965_2.fastq.gz SRR10512965.ctx\n",
      "+ set +x\n",
      "Working on SRR10512968\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s SRR10512968 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10512968_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10512968_2.fastq.gz SRR10512968.ctx\n",
      "+ set +x\n",
      "Working on SRR10513325\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s SRR10513325 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10513325_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10513325_2.fastq.gz SRR10513325.ctx\n",
      "+ set +x\n",
      "Working on SRR10513326\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s SRR10513326 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10513326_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10513326_2.fastq.gz SRR10513326.ctx\n",
      "+ set +x\n",
      "Working on SRR10513328\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s SRR10513328 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10513328_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10513328_2.fastq.gz SRR10513328.ctx\n",
      "+ set +x\n",
      "Working on SRR10513363\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s SRR10513363 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10513363_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10513363_2.fastq.gz SRR10513363.ctx\n",
      "+ set +x\n",
      "Working on SRR10513672\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s SRR10513672 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10513672_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10513672_2.fastq.gz SRR10513672.ctx\n",
      "+ set +x\n",
      "Working on SRR10519468\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s SRR10519468 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10519468_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10519468_2.fastq.gz SRR10519468.ctx\n",
      "+ set +x\n",
      "Working on SRR10519469\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s SRR10519469 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10519469_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10519469_2.fastq.gz SRR10519469.ctx\n",
      "+ set +x\n",
      "Working on SRR10519616\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s SRR10519616 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10519616_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10519616_2.fastq.gz SRR10519616.ctx\n",
      "+ set +x\n",
      "Working on SRR10519617\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s SRR10519617 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10519617_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10519617_2.fastq.gz SRR10519617.ctx\n",
      "+ set +x\n",
      "Working on SRR10519619\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s SRR10519619 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10519619_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10519619_2.fastq.gz SRR10519619.ctx\n",
      "+ set +x\n",
      "Working on SRR10519620\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s SRR10519620 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10519620_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10519620_2.fastq.gz SRR10519620.ctx\n",
      "+ set +x\n",
      "Working on SRR10519637\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s SRR10519637 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10519637_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10519637_2.fastq.gz SRR10519637.ctx\n",
      "+ set +x\n",
      "Working on SRR10521982\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s SRR10521982 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10521982_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10521982_2.fastq.gz SRR10521982.ctx\n",
      "+ set +x\n",
      "Working on SRR10521983\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s SRR10521983 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10521983_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10521983_2.fastq.gz SRR10521983.ctx\n",
      "+ set +x\n",
      "Working on SRR10521984\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s SRR10521984 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10521984_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10521984_2.fastq.gz SRR10521984.ctx\n",
      "+ set +x\n",
      "Working on SRR10527348\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s SRR10527348 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10527348_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10527348_2.fastq.gz SRR10527348.ctx\n",
      "+ set +x\n",
      "Working on SRR10527349\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s SRR10527349 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10527349_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10527349_2.fastq.gz SRR10527349.ctx\n",
      "+ set +x\n",
      "Working on SRR10527351\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s SRR10527351 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10527351_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10527351_2.fastq.gz SRR10527351.ctx\n",
      "+ set +x\n",
      "Working on SRR10527352\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s SRR10527352 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10527352_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10527352_2.fastq.gz SRR10527352.ctx\n",
      "+ set +x\n",
      "Working on SRR10527353\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s SRR10527353 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10527353_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR10527353_2.fastq.gz SRR10527353.ctx\n",
      "+ set +x\n",
      "Working on SRR8088181\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s SRR8088181 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR8088181_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR8088181_2.fastq.gz SRR8088181.ctx\n",
      "+ set +x\n",
      "Working on SRR8088182\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s SRR8088182 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR8088182_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR8088182_2.fastq.gz SRR8088182.ctx\n",
      "+ set +x\n",
      "Working on SRR8088183\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s SRR8088183 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR8088183_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR8088183_2.fastq.gz SRR8088183.ctx\n",
      "+ set +x\n",
      "Working on SRR8088184\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s SRR8088184 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR8088184_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR8088184_2.fastq.gz SRR8088184.ctx\n",
      "+ set +x\n",
      "Working on SRR8088185\n",
      "+ conda run --name mccortex mccortex 9 build -t 32 -k 9 -s SRR8088185 --seq2 /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR8088185_1.fastq.gz:/home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/data/SRR8088185_2.fastq.gz SRR8088185.ctx\n",
      "+ set +x\n",
      "Done\n"
     ]
    }
   ],
   "source": [
    "cd $ROOT_DIR\n",
    "mkdir bigsi\n",
    "pushd bigsi\n",
    "\n",
    "for genome in $ROOT_DIR/data/*_1.fastq.gz;\n",
    "do\n",
    "    accession=`basename $genome _1.fastq.gz`\n",
    "    echo \"Working on $accession\"\n",
    "       \n",
    "    fastq_file_1=$ROOT_DIR/data/${accession}_1.fastq.gz\n",
    "    fastq_file_2=$ROOT_DIR/data/${accession}_2.fastq.gz\n",
    "    cortex_out=${accession}.ctx\n",
    "    cortex_log=${accession}_mccortex.log\n",
    "    \n",
    "    set -x\n",
    "    conda run --name mccortex mccortex ${kmer_size} build -t ${threads} -k ${kmer_size} -s ${accession} --seq2 ${fastq_file_1}:${fastq_file_2} ${cortex_out} 2> ${cortex_log}.err 1> ${cortex_log}\n",
    "    set +x\n",
    "done\n",
    "\n",
    "echo \"Done\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ERR1144974.ctx\n",
      "ERR1144975.ctx\n",
      "ERR1144976.ctx\n",
      "ERR1144977.ctx\n",
      "ERR1144978.ctx\n",
      "ERR3655992.ctx\n",
      "ERR3655994.ctx\n",
      "ERR3655996.ctx\n",
      "ERR3655998.ctx\n",
      "ERR3656002.ctx\n",
      "50\n",
      "82M\t.\n"
     ]
    }
   ],
   "source": [
    "ls *.ctx|head\n",
    "ls *.ctx | wc -l\n",
    "du -sh ."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Awesome. We've now got files named like `SRR8088185.ctx` in our directory containing the cortex graph/kmers."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 2: Build BIGSI index\n",
    "\n",
    "In step 2, we'll look at building the `bigsi` indexes for each of these kmer counts.\n",
    "\n",
    "Let's setup the bigsi index configuration."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "cat > berkeleydb.yaml << EOF\n",
    "## Example config using berkeleyDB\n",
    "h: ${number_hashes}\n",
    "k: ${kmer_size}\n",
    "m: ${m_value}\n",
    "storage-engine: berkeleydb\n",
    "storage-config:\n",
    "  filename: bigsi.db\n",
    "  flag: \"c\" ## Change to 'r' for read-only access\n",
    "EOF\n",
    "\n",
    "export BIGSI_CONFIG=berkeleydb.yaml"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, let's construct the blooom filters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bigsi-v0.3.1\n",
      "\n",
      "Available Commands:\n",
      "\n",
      " - bloom: Creates a bloom filter from a sequence file or cortex graph. (fastq...\n",
      " - build\n",
      " - delete\n",
      " - insert: Inserts a bloom filter into the graph          e.g. bigsi insert E...\n",
      " - merge\n",
      " - search\n",
      "\n",
      "ERROR conda.cli.main_run:execute(39): Subprocess for 'conda run ['bigsi']' command failed.  Stderr was:\n",
      "\n",
      "+ parallel --jobs 32 -I% conda run --name bigsi bigsi bloom % %.bloom ::: ERR1144974.ctx ERR1144975.ctx ERR1144976.ctx ERR1144977.ctx ERR1144978.ctx ERR3655992.ctx ERR3655994.ctx ERR3655996.ctx ERR3655998.ctx ERR3656002.ctx ERR3656004.ctx ERR3656010.ctx ERR3656012.ctx ERR3656013.ctx ERR3656015.ctx ERR3656018.ctx ERR3656019.ctx SRR10298903.ctx SRR10298904.ctx SRR10298905.ctx SRR10298906.ctx SRR10298907.ctx SRR10512964.ctx SRR10512965.ctx SRR10512968.ctx SRR10513325.ctx SRR10513326.ctx SRR10513328.ctx SRR10513363.ctx SRR10513672.ctx SRR10519468.ctx SRR10519469.ctx SRR10519616.ctx SRR10519617.ctx SRR10519619.ctx SRR10519620.ctx SRR10519637.ctx SRR10521982.ctx SRR10521983.ctx SRR10521984.ctx SRR10527348.ctx SRR10527349.ctx SRR10527351.ctx SRR10527352.ctx SRR10527353.ctx SRR8088181.ctx SRR8088182.ctx SRR8088183.ctx SRR8088184.ctx SRR8088185.ctx\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "+ set +x\n"
     ]
    }
   ],
   "source": [
    "conda run --name bigsi bigsi\n",
    "set -x\n",
    "parallel --jobs ${threads} -I% conda run --name bigsi bigsi bloom % %.bloom ::: *.ctx\n",
    "set +x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ERR1144974.ctx.bloom\n",
      "ERR1144975.ctx.bloom\n",
      "ERR1144976.ctx.bloom\n",
      "ERR1144977.ctx.bloom\n",
      "ERR1144978.ctx.bloom\n",
      "ERR3655992.ctx.bloom\n",
      "ERR3655994.ctx.bloom\n",
      "ERR3655996.ctx.bloom\n",
      "ERR3655998.ctx.bloom\n",
      "ERR3656002.ctx.bloom\n",
      "50\n"
     ]
    }
   ],
   "source": [
    "ls -d *.bloom|head\n",
    "ls -d *.bloom|wc -l"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Awesome. We now have our BIGSI bloom filters.\n",
    "\n",
    "Let's merge these all into a BIGSI index."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+ conda run --name bigsi bigsi build ERR1144974.ctx.bloom ERR1144975.ctx.bloom ERR1144976.ctx.bloom ERR1144977.ctx.bloom ERR1144978.ctx.bloom ERR3655992.ctx.bloom ERR3655994.ctx.bloom ERR3655996.ctx.bloom ERR3655998.ctx.bloom ERR3656002.ctx.bloom ERR3656004.ctx.bloom ERR3656010.ctx.bloom ERR3656012.ctx.bloom ERR3656013.ctx.bloom ERR3656015.ctx.bloom ERR3656018.ctx.bloom ERR3656019.ctx.bloom SRR10298903.ctx.bloom SRR10298904.ctx.bloom SRR10298905.ctx.bloom SRR10298906.ctx.bloom SRR10298907.ctx.bloom SRR10512964.ctx.bloom SRR10512965.ctx.bloom SRR10512968.ctx.bloom SRR10513325.ctx.bloom SRR10513326.ctx.bloom SRR10513328.ctx.bloom SRR10513363.ctx.bloom SRR10513672.ctx.bloom SRR10519468.ctx.bloom SRR10519469.ctx.bloom SRR10519616.ctx.bloom SRR10519617.ctx.bloom SRR10519619.ctx.bloom SRR10519620.ctx.bloom SRR10519637.ctx.bloom SRR10521982.ctx.bloom SRR10521983.ctx.bloom SRR10521984.ctx.bloom SRR10527348.ctx.bloom SRR10527349.ctx.bloom SRR10527351.ctx.bloom SRR10527352.ctx.bloom SRR10527353.ctx.bloom SRR8088181.ctx.bloom SRR8088182.ctx.bloom SRR8088183.ctx.bloom SRR8088184.ctx.bloom SRR8088185.ctx.bloom -s ERR1144974 -s ERR1144975 -s ERR1144976 -s ERR1144977 -s ERR1144978 -s ERR3655992 -s ERR3655994 -s ERR3655996 -s ERR3655998 -s ERR3656002 -s ERR3656004 -s ERR3656010 -s ERR3656012 -s ERR3656013 -s ERR3656015 -s ERR3656018 -s ERR3656019 -s SRR10298903 -s SRR10298904 -s SRR10298905 -s SRR10298906 -s SRR10298907 -s SRR10512964 -s SRR10512965 -s SRR10512968 -s SRR10513325 -s SRR10513326 -s SRR10513328 -s SRR10513363 -s SRR10513672 -s SRR10519468 -s SRR10519469 -s SRR10519616 -s SRR10519617 -s SRR10519619 -s SRR10519620 -s SRR10519637 -s SRR10521982 -s SRR10521983 -s SRR10521984 -s SRR10527348 -s SRR10527349 -s SRR10527351 -s SRR10527352 -s SRR10527353 -s SRR8088181 -s SRR8088182 -s SRR8088183 -s SRR8088184 -s SRR8088185\n",
      "+ '[' 155 -lt 1 ']'\n",
      "+ local cmd=run\n",
      "+ shift\n",
      "+ case \"$cmd\" in\n",
      "+ OLDPATH=/home/CSCScience.ca/apetkau/.local/bin:/home/CSCScience.ca/apetkau/miniconda3/bin:/home/CSCScience.ca/apetkau/miniconda3/condabin:/home/CSCScience.ca/apetkau/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/opt/puppetlabs/bin\n",
      "+ __add_sys_prefix_to_path\n",
      "+ '[' -n '' ']'\n",
      "++ dirname /home/CSCScience.ca/apetkau/miniconda3/bin/conda\n",
      "+ SYSP=/home/CSCScience.ca/apetkau/miniconda3/bin\n",
      "++ dirname /home/CSCScience.ca/apetkau/miniconda3/bin\n",
      "+ SYSP=/home/CSCScience.ca/apetkau/miniconda3\n",
      "+ '[' -n '' ']'\n",
      "+ PATH=/home/CSCScience.ca/apetkau/miniconda3/bin:/home/CSCScience.ca/apetkau/.local/bin:/home/CSCScience.ca/apetkau/miniconda3/bin:/home/CSCScience.ca/apetkau/miniconda3/condabin:/home/CSCScience.ca/apetkau/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/opt/puppetlabs/bin\n",
      "+ export PATH\n",
      "+ /home/CSCScience.ca/apetkau/miniconda3/bin/conda run --name bigsi bigsi build ERR1144974.ctx.bloom ERR1144975.ctx.bloom ERR1144976.ctx.bloom ERR1144977.ctx.bloom ERR1144978.ctx.bloom ERR3655992.ctx.bloom ERR3655994.ctx.bloom ERR3655996.ctx.bloom ERR3655998.ctx.bloom ERR3656002.ctx.bloom ERR3656004.ctx.bloom ERR3656010.ctx.bloom ERR3656012.ctx.bloom ERR3656013.ctx.bloom ERR3656015.ctx.bloom ERR3656018.ctx.bloom ERR3656019.ctx.bloom SRR10298903.ctx.bloom SRR10298904.ctx.bloom SRR10298905.ctx.bloom SRR10298906.ctx.bloom SRR10298907.ctx.bloom SRR10512964.ctx.bloom SRR10512965.ctx.bloom SRR10512968.ctx.bloom SRR10513325.ctx.bloom SRR10513326.ctx.bloom SRR10513328.ctx.bloom SRR10513363.ctx.bloom SRR10513672.ctx.bloom SRR10519468.ctx.bloom SRR10519469.ctx.bloom SRR10519616.ctx.bloom SRR10519617.ctx.bloom SRR10519619.ctx.bloom SRR10519620.ctx.bloom SRR10519637.ctx.bloom SRR10521982.ctx.bloom SRR10521983.ctx.bloom SRR10521984.ctx.bloom SRR10527348.ctx.bloom SRR10527349.ctx.bloom SRR10527351.ctx.bloom SRR10527352.ctx.bloom SRR10527353.ctx.bloom SRR8088181.ctx.bloom SRR8088182.ctx.bloom SRR8088183.ctx.bloom SRR8088184.ctx.bloom SRR8088185.ctx.bloom -s ERR1144974 -s ERR1144975 -s ERR1144976 -s ERR1144977 -s ERR1144978 -s ERR3655992 -s ERR3655994 -s ERR3655996 -s ERR3655998 -s ERR3656002 -s ERR3656004 -s ERR3656010 -s ERR3656012 -s ERR3656013 -s ERR3656015 -s ERR3656018 -s ERR3656019 -s SRR10298903 -s SRR10298904 -s SRR10298905 -s SRR10298906 -s SRR10298907 -s SRR10512964 -s SRR10512965 -s SRR10512968 -s SRR10513325 -s SRR10513326 -s SRR10513328 -s SRR10513363 -s SRR10513672 -s SRR10519468 -s SRR10519469 -s SRR10519616 -s SRR10519617 -s SRR10519619 -s SRR10519620 -s SRR10519637 -s SRR10521982 -s SRR10521983 -s SRR10521984 -s SRR10527348 -s SRR10527349 -s SRR10527351 -s SRR10527352 -s SRR10527353 -s SRR8088181 -s SRR8088182 -s SRR8088183 -s SRR8088184 -s SRR8088185\n",
      "{'result': 'success'}\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/bigsi/lib/python3.7/site-packages/bigsi/__main__.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  config = yaml.load(infile)\n",
      "INFO:bigsi.cmds.build:Building index: 0/1\n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/ERR1144974.ctx.bloom/ERR1144974.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/ERR1144975.ctx.bloom/ERR1144975.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/ERR1144976.ctx.bloom/ERR1144976.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/ERR1144977.ctx.bloom/ERR1144977.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/ERR1144978.ctx.bloom/ERR1144978.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/ERR3655992.ctx.bloom/ERR3655992.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/ERR3655994.ctx.bloom/ERR3655994.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/ERR3655996.ctx.bloom/ERR3655996.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/ERR3655998.ctx.bloom/ERR3655998.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/ERR3656002.ctx.bloom/ERR3656002.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/ERR3656004.ctx.bloom/ERR3656004.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/ERR3656010.ctx.bloom/ERR3656010.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/ERR3656012.ctx.bloom/ERR3656012.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/ERR3656013.ctx.bloom/ERR3656013.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/ERR3656015.ctx.bloom/ERR3656015.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/ERR3656018.ctx.bloom/ERR3656018.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/ERR3656019.ctx.bloom/ERR3656019.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/SRR10298903.ctx.bloom/SRR10298903.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/SRR10298904.ctx.bloom/SRR10298904.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/SRR10298905.ctx.bloom/SRR10298905.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/SRR10298906.ctx.bloom/SRR10298906.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/SRR10298907.ctx.bloom/SRR10298907.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/SRR10512964.ctx.bloom/SRR10512964.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/SRR10512965.ctx.bloom/SRR10512965.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/SRR10512968.ctx.bloom/SRR10512968.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/SRR10513325.ctx.bloom/SRR10513325.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/SRR10513326.ctx.bloom/SRR10513326.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/SRR10513328.ctx.bloom/SRR10513328.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/SRR10513363.ctx.bloom/SRR10513363.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/SRR10513672.ctx.bloom/SRR10513672.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/SRR10519468.ctx.bloom/SRR10519468.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/SRR10519469.ctx.bloom/SRR10519469.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/SRR10519616.ctx.bloom/SRR10519616.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/SRR10519617.ctx.bloom/SRR10519617.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/SRR10519619.ctx.bloom/SRR10519619.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/SRR10519620.ctx.bloom/SRR10519620.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/SRR10519637.ctx.bloom/SRR10519637.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/SRR10521982.ctx.bloom/SRR10521982.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/SRR10521983.ctx.bloom/SRR10521983.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/SRR10521984.ctx.bloom/SRR10521984.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/SRR10527348.ctx.bloom/SRR10527348.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/SRR10527349.ctx.bloom/SRR10527349.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/SRR10527351.ctx.bloom/SRR10527351.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/SRR10527352.ctx.bloom/SRR10527352.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/SRR10527353.ctx.bloom/SRR10527353.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/SRR8088181.ctx.bloom/SRR8088181.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/SRR8088182.ctx.bloom/SRR8088182.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/SRR8088183.ctx.bloom/SRR8088183.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/SRR8088184.ctx.bloom/SRR8088184.ctx.bloom \n",
      "DEBUG:bigsi.cmds.build:Loading /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/bigsi/SRR8088185.ctx.bloom/SRR8088185.ctx.bloom \n",
      "DEBUG:bigsi.graph.bigsi:Insert sample metadata\n",
      "DEBUG:bigsi.graph.bigsi:Create signature index\n",
      "DEBUG:bigsi.graph.index:Transpose bitarrays\n",
      "DEBUG:bigsi.graph.index:Insert rows\n",
      "DEBUG:bigsi.storage.base:set bitarrays\n",
      "+ local t1=0\n",
      "+ PATH=/home/CSCScience.ca/apetkau/.local/bin:/home/CSCScience.ca/apetkau/miniconda3/bin:/home/CSCScience.ca/apetkau/miniconda3/condabin:/home/CSCScience.ca/apetkau/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/opt/puppetlabs/bin\n",
      "+ return 0\n",
      "\n",
      "real\t6m3.136s\n",
      "user\t4m4.009s\n",
      "sys\t1m59.880s\n",
      "+ set +x\n"
     ]
    }
   ],
   "source": [
    "# Builds command-line string\n",
    "samples=`for i in *.bloom; do echo -n \"-s \"; basename $i .ctx.bloom; done`\n",
    "\n",
    "set -x\n",
    "time conda run --name bigsi bigsi build *.bloom ${samples}\n",
    "set +x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-r--r-- 1 apetkau grp_apetkau 1.3G Dec  3 12:05 bigsi.db\n"
     ]
    }
   ],
   "source": [
    "ls -lh bigsi.db"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Awesome. We've gotten our database constructed. Let's try it out."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sample_name\tpercent_kmers\n",
      "ERR1144974\t100\n",
      "ERR1144975\t100\n",
      "ERR1144976\t100\n",
      "ERR1144977\t100\n",
      "ERR1144978\t100\n",
      "ERR3655992\t100\n",
      "ERR3655994\t100\n",
      "\n",
      "real\t0m1.310s\n",
      "user\t0m3.438s\n",
      "sys\t0m3.947s\n"
     ]
    }
   ],
   "source": [
    "echo -e \"sample_name\\tpercent_kmers\"\n",
    "time conda run --name bigsi bigsi search GTTTCGTTCTTCCGGCGCGGGCGGTCAGCACGTTAACACCACCGACTCCGCTATCCGTATTACCCACTTGCCGACCGGCATCTTGGTGGAATGCCAGGACGAGC 2> /dev/null \\\n",
    "    | sed -e $'s/\\'/\\\"/g' | jq -r '.results[] | \"\\(.sample_name)\\t\\(.percent_kmers_found)\"'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Awesome. We can use this BIGSI index to pull out sample identifiers with specific genes."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
