{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "all_kat_hist                            \u001b[0m\u001b[01;34mimages\u001b[0m\n",
      "all_kat_hist.dist_analysis.json         kat.hist\n",
      "\u001b[01;35mall_kat_hist.png\u001b[0m                        kat.hist.dist_analysis.json\n",
      "\u001b[01;34mbigsi\u001b[0m                                   \u001b[01;35mkat.hist.png\u001b[0m\n",
      "\u001b[01;34mbigsi.bak\u001b[0m                               microbial-assembly.ipynb\n",
      "\u001b[01;34mdata\u001b[0m                                    microbial-bigsi.ipynb\n",
      "ERR1144974                              microbial-genomes.txt\n",
      "ERR1144974.dist_analysis.json           microbial-process-data.ipynb\n",
      "ERR1144974_kat_hist                     microbial-process-data.ipynb.bak\n",
      "ERR1144974_kat_hist.dist_analysis.json  \u001b[01;35mo.png\u001b[0m\n",
      "\u001b[01;35mERR1144974_kat_hist.png\u001b[0m                 sha256\n",
      "\u001b[01;35mERR1144974.png\u001b[0m                          total-bp-reads.txt\n",
      "freq_k31.hist                           total-kmers.txt\n",
      "freq_k7.hist\n"
     ]
    }
   ],
   "source": [
    "NOTEBOOK_DIR=`git rev-parse --show-toplevel`\n",
    "ROOT_DIR=$NOTEBOOK_DIR/microbial\n",
    "cd $ROOT_DIR\n",
    "ls"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 1: Assemble genomes\n",
    "\n",
    "Let's try assembling the genomes using [skesa](https://github.com/ncbi/SKESA)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "threads=48"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘assembly’: File exists\n",
      "SKESA v.2.3.0\n",
      "skesa --version \n",
      "\n",
      "ERR1144974 ... ERR1144975 ... ERR1144976 ... ERR1144977 ... ERR1144978 ... ERR3655992 ... ERR3655994 ... ERR3655996 ... ERR3655998 ... ERR3656002 ... ERR3656004 ... ERR3656010 ... ERR3656012 ... ERR3656013 ... ERR3656015 ... ERR3656018 ... ERR3656019 ... SRR10298903 ... SRR10298904 ... SRR10298905 ... SRR10298906 ... SRR10298907 ... SRR10512964 ... SRR10512965 ... SRR10512968 ... SRR10513325 ... SRR10513326 ... SRR10513328 ... SRR10513363 ... SRR10513672 ... SRR10519468 ... SRR10519469 ... SRR10519616 ... SRR10519617 ... SRR10519619 ... SRR10519620 ... SRR10519637 ... SRR10521982 ... SRR10521983 ... SRR10521984 ... SRR10527348 ... SRR10527349 ... SRR10527351 ... SRR10527352 ... SRR10527353 ... SRR8088181 ... SRR8088182 ... SRR8088183 ... SRR8088184 ... SRR8088185 ... done\n"
     ]
    }
   ],
   "source": [
    "mkdir assembly\n",
    "\n",
    "conda run --name skesa skesa --version\n",
    "\n",
    "data_dir=data/subsample\n",
    "for file in ${data_dir}/*_1.fastq.gz\n",
    "do\n",
    "    name=`basename $file _1.fastq.gz`\n",
    "    \n",
    "    input1=${data_dir}/${name}_1.fastq.gz\n",
    "    input2=${data_dir}/${name}_2.fastq.gz\n",
    "    \n",
    "    output=assembly/${name}.fasta\n",
    "    log=assembly/${name}.log\n",
    "    \n",
    "    echo \"${name}\"\n",
    "    \n",
    "    conda run --name skesa /usr/bin/time -v -o assembly/${name}.time \\\n",
    "        skesa --cores ${threads} --fastq ${input1},${input2} --contigs_out ${output} 1> ${log}.out 2> ${log}.err\n",
    "done\n",
    "echo \"Done\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-r--r-- 1 apetkau grp_apetkau 4.2M Dec  4 17:47 assembly/ERR1144974.fasta\n",
      "-rw-r--r-- 1 apetkau grp_apetkau 4.2M Dec  4 17:48 assembly/ERR1144975.fasta\n",
      "-rw-r--r-- 1 apetkau grp_apetkau 4.1M Dec  4 17:48 assembly/ERR1144976.fasta\n",
      "-rw-r--r-- 1 apetkau grp_apetkau 4.1M Dec  4 17:49 assembly/ERR1144977.fasta\n",
      "-rw-r--r-- 1 apetkau grp_apetkau 4.1M Dec  4 17:50 assembly/ERR1144978.fasta\n",
      "-rw-r--r-- 1 apetkau grp_apetkau 6.7M Dec  4 17:52 assembly/ERR3655992.fasta\n",
      "-rw-r--r-- 1 apetkau grp_apetkau 6.8M Dec  4 17:54 assembly/ERR3655994.fasta\n",
      "-rw-r--r-- 1 apetkau grp_apetkau 6.2M Dec  4 17:56 assembly/ERR3655996.fasta\n",
      "-rw-r--r-- 1 apetkau grp_apetkau 6.3M Dec  4 17:58 assembly/ERR3655998.fasta\n",
      "-rw-r--r-- 1 apetkau grp_apetkau 6.4M Dec  4 18:00 assembly/ERR3656002.fasta\n",
      "50\n"
     ]
    }
   ],
   "source": [
    "ls -lh assembly/*.fasta | head\n",
    "ls assembly/*.fasta | wc -l"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Awesome. We've got assembled genomes.\n",
    "\n",
    "# Step 2: Evaluate quality\n",
    "\n",
    "Let's evaluate the quality using [quast](http://bioinf.spbau.ru/quast)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/CSCScience.ca/apetkau/miniconda3/envs/quast/bin/quast -o assembly/quast assembly/ERR1144974.fasta assembly/ERR1144975.fasta assembly/ERR1144976.fasta assembly/ERR1144977.fasta assembly/ERR1144978.fasta assembly/ERR3655992.fasta assembly/ERR3655994.fasta assembly/ERR3655996.fasta assembly/ERR3655998.fasta assembly/ERR3656002.fasta assembly/ERR3656004.fasta assembly/ERR3656010.fasta assembly/ERR3656012.fasta assembly/ERR3656013.fasta assembly/ERR3656015.fasta assembly/ERR3656018.fasta assembly/ERR3656019.fasta assembly/SRR10298903.fasta assembly/SRR10298904.fasta assembly/SRR10298905.fasta assembly/SRR10298906.fasta assembly/SRR10298907.fasta assembly/SRR10512964.fasta assembly/SRR10512965.fasta assembly/SRR10512968.fasta assembly/SRR10513325.fasta assembly/SRR10513326.fasta assembly/SRR10513328.fasta assembly/SRR10513363.fasta assembly/SRR10513672.fasta assembly/SRR10519468.fasta assembly/SRR10519469.fasta assembly/SRR10519616.fasta assembly/SRR10519617.fasta assembly/SRR10519619.fasta assembly/SRR10519620.fasta assembly/SRR10519637.fasta assembly/SRR10521982.fasta assembly/SRR10521983.fasta assembly/SRR10521984.fasta assembly/SRR10527348.fasta assembly/SRR10527349.fasta assembly/SRR10527351.fasta assembly/SRR10527352.fasta assembly/SRR10527353.fasta assembly/SRR8088181.fasta assembly/SRR8088182.fasta assembly/SRR8088183.fasta assembly/SRR8088184.fasta assembly/SRR8088185.fasta\n",
      "\n",
      "Version: 5.0.2\n",
      "\n",
      "System information:\n",
      "  OS: Linux-4.15.0-70-generic-x86_64-with-debian-buster-sid (linux_64)\n",
      "  Python version: 3.6.7\n",
      "  CPUs number: 56\n",
      "\n",
      "Started: 2019-12-04 18:53:59\n",
      "\n",
      "Logging to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/quast.log\n",
      "NOTICE: Maximum number of threads is set to 14 (use --threads option to set it manually)\n",
      "\n",
      "CWD: /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial\n",
      "Main parameters: \n",
      "  MODE: default, threads: 14, minimum contig length: 500, minimum alignment length: 65, \\\n",
      "  ambiguity: one, threshold for extensive misassembly size: 1000\n",
      "\n",
      "Contigs:\n",
      "  Pre-processing...\n",
      "  1  assembly/ERR1144974.fasta ==> ERR1144974\n",
      "  2  assembly/ERR1144975.fasta ==> ERR1144975\n",
      "  3  assembly/ERR1144976.fasta ==> ERR1144976\n",
      "  4  assembly/ERR1144977.fasta ==> ERR1144977\n",
      "  5  assembly/ERR1144978.fasta ==> ERR1144978\n",
      "  6  assembly/ERR3655992.fasta ==> ERR3655992\n",
      "  7  assembly/ERR3655994.fasta ==> ERR3655994\n",
      "  8  assembly/ERR3655996.fasta ==> ERR3655996\n",
      "  9  assembly/ERR3655998.fasta ==> ERR3655998\n",
      "  10 assembly/ERR3656002.fasta ==> ERR3656002\n",
      "  11 assembly/ERR3656004.fasta ==> ERR3656004\n",
      "  12 assembly/ERR3656010.fasta ==> ERR3656010\n",
      "  13 assembly/ERR3656012.fasta ==> ERR3656012\n",
      "  14 assembly/ERR3656013.fasta ==> ERR3656013\n",
      "  15 assembly/ERR3656015.fasta ==> ERR3656015\n",
      "  16 assembly/ERR3656018.fasta ==> ERR3656018\n",
      "  17 assembly/ERR3656019.fasta ==> ERR3656019\n",
      "  18 assembly/SRR10298903.fasta ==> SRR10298903\n",
      "  19 assembly/SRR10298904.fasta ==> SRR10298904\n",
      "  20 assembly/SRR10298905.fasta ==> SRR10298905\n",
      "  21 assembly/SRR10298906.fasta ==> SRR10298906\n",
      "  22 assembly/SRR10298907.fasta ==> SRR10298907\n",
      "  23 assembly/SRR10512964.fasta ==> SRR10512964\n",
      "  24 assembly/SRR10512965.fasta ==> SRR10512965\n",
      "  25 assembly/SRR10512968.fasta ==> SRR10512968\n",
      "  26 assembly/SRR10513325.fasta ==> SRR10513325\n",
      "  27 assembly/SRR10513326.fasta ==> SRR10513326\n",
      "  28 assembly/SRR10513328.fasta ==> SRR10513328\n",
      "  29 assembly/SRR10513363.fasta ==> SRR10513363\n",
      "  30 assembly/SRR10513672.fasta ==> SRR10513672\n",
      "  31 assembly/SRR10519468.fasta ==> SRR10519468\n",
      "  32 assembly/SRR10519469.fasta ==> SRR10519469\n",
      "  33 assembly/SRR10519616.fasta ==> SRR10519616\n",
      "  34 assembly/SRR10519617.fasta ==> SRR10519617\n",
      "  35 assembly/SRR10519619.fasta ==> SRR10519619\n",
      "  36 assembly/SRR10519620.fasta ==> SRR10519620\n",
      "  37 assembly/SRR10519637.fasta ==> SRR10519637\n",
      "  38 assembly/SRR10521982.fasta ==> SRR10521982\n",
      "  39 assembly/SRR10521983.fasta ==> SRR10521983\n",
      "  40 assembly/SRR10521984.fasta ==> SRR10521984\n",
      "  41 assembly/SRR10527348.fasta ==> SRR10527348\n",
      "  42 assembly/SRR10527349.fasta ==> SRR10527349\n",
      "  43 assembly/SRR10527351.fasta ==> SRR10527351\n",
      "  44 assembly/SRR10527352.fasta ==> SRR10527352\n",
      "  45 assembly/SRR10527353.fasta ==> SRR10527353\n",
      "  46 assembly/SRR8088181.fasta ==> SRR8088181\n",
      "  47 assembly/SRR8088182.fasta ==> SRR8088182\n",
      "  48 assembly/SRR8088183.fasta ==> SRR8088183\n",
      "  49 assembly/SRR8088184.fasta ==> SRR8088184\n",
      "  50 assembly/SRR8088185.fasta ==> SRR8088185\n",
      "\n",
      "2019-12-04 18:54:00\n",
      "Running Basic statistics processor...\n",
      "  Contig files: \n",
      "    1  ERR1144974\n",
      "    2  ERR1144975\n",
      "    3  ERR1144976\n",
      "    4  ERR1144977\n",
      "    5  ERR1144978\n",
      "    6  ERR3655992\n",
      "    7  ERR3655994\n",
      "    8  ERR3655996\n",
      "    9  ERR3655998\n",
      "    10 ERR3656002\n",
      "    11 ERR3656004\n",
      "    12 ERR3656010\n",
      "    13 ERR3656012\n",
      "    14 ERR3656013\n",
      "    15 ERR3656015\n",
      "    16 ERR3656018\n",
      "    17 ERR3656019\n",
      "    18 SRR10298903\n",
      "    19 SRR10298904\n",
      "    20 SRR10298905\n",
      "    21 SRR10298906\n",
      "    22 SRR10298907\n",
      "    23 SRR10512964\n",
      "    24 SRR10512965\n",
      "    25 SRR10512968\n",
      "    26 SRR10513325\n",
      "    27 SRR10513326\n",
      "    28 SRR10513328\n",
      "    29 SRR10513363\n",
      "    30 SRR10513672\n",
      "    31 SRR10519468\n",
      "    32 SRR10519469\n",
      "    33 SRR10519616\n",
      "    34 SRR10519617\n",
      "    35 SRR10519619\n",
      "    36 SRR10519620\n",
      "    37 SRR10519637\n",
      "    38 SRR10521982\n",
      "    39 SRR10521983\n",
      "    40 SRR10521984\n",
      "    41 SRR10527348\n",
      "    42 SRR10527349\n",
      "    43 SRR10527351\n",
      "    44 SRR10527352\n",
      "    45 SRR10527353\n",
      "    46 SRR8088181\n",
      "    47 SRR8088182\n",
      "    48 SRR8088183\n",
      "    49 SRR8088184\n",
      "    50 SRR8088185\n",
      "  Calculating N50 and L50...\n",
      "    1  ERR1144974, N50 = 62596, L50 = 23, Total length = 4302248, GC % = 65.48, # N's per 100 kbp =  0.00\n",
      "    2  ERR1144975, N50 = 49776, L50 = 26, Total length = 4282413, GC % = 65.37, # N's per 100 kbp =  0.00\n",
      "    3  ERR1144976, N50 = 58119, L50 = 25, Total length = 4274859, GC % = 65.35, # N's per 100 kbp =  0.00\n",
      "    4  ERR1144977, N50 = 54933, L50 = 25, Total length = 4278749, GC % = 65.35, # N's per 100 kbp =  0.00\n",
      "    5  ERR1144978, N50 = 57167, L50 = 24, Total length = 4277560, GC % = 65.36, # N's per 100 kbp =  0.00\n",
      "    6  ERR3655992, N50 = 204705, L50 = 12, Total length = 6973935, GC % = 65.87, # N's per 100 kbp =  0.00\n",
      "    7  ERR3655994, N50 = 196809, L50 = 14, Total length = 7022163, GC % = 65.80, # N's per 100 kbp =  0.00\n",
      "    8  ERR3655996, N50 = 396051, L50 = 6, Total length = 6465887, GC % = 66.25, # N's per 100 kbp =  0.00\n",
      "    9  ERR3655998, N50 = 338360, L50 = 6, Total length = 6564366, GC % = 66.26, # N's per 100 kbp =  0.00\n",
      "    10 ERR3656002, N50 = 323429, L50 = 8, Total length = 6688960, GC % = 65.98, # N's per 100 kbp =  0.00\n",
      "    11 ERR3656004, N50 = 486845, L50 = 2, Total length = 3765674, GC % = 38.92, # N's per 100 kbp =  0.00\n",
      "    12 ERR3656010, N50 = 103737, L50 = 13, Total length = 3939559, GC % = 38.83, # N's per 100 kbp =  0.00\n",
      "    13 ERR3656012, N50 = 103142, L50 = 13, Total length = 3981026, GC % = 38.86, # N's per 100 kbp =  0.00\n",
      "    14 ERR3656013, N50 = 109955, L50 = 12, Total length = 3982476, GC % = 38.86, # N's per 100 kbp =  0.00\n",
      "    15 ERR3656015, N50 = 103142, L50 = 13, Total length = 3979770, GC % = 38.86, # N's per 100 kbp =  0.00\n",
      "    16 ERR3656018, N50 = 166041, L50 = 12, Total length = 5645710, GC % = 56.95, # N's per 100 kbp =  0.00\n",
      "    17 ERR3656019, N50 = 214399, L50 = 8, Total length = 5718407, GC % = 56.75, # N's per 100 kbp =  0.00\n",
      "    18 SRR10298903, N50 = 162539, L50 = 8, Total length = 3436529, GC % = 38.23, # N's per 100 kbp =  0.00\n",
      "    19 SRR10298904, N50 = 276473, L50 = 4, Total length = 3437596, GC % = 38.22, # N's per 100 kbp =  0.00\n",
      "    20 SRR10298905, N50 = 98871, L50 = 12, Total length = 3432037, GC % = 38.24, # N's per 100 kbp =  0.00\n",
      "    21 SRR10298906, N50 = 218915, L50 = 7, Total length = 3436620, GC % = 38.23, # N's per 100 kbp =  0.00\n",
      "    22 SRR10298907, N50 = 106809, L50 = 11, Total length = 3639675, GC % = 38.23, # N's per 100 kbp =  0.00\n",
      "    23 SRR10512964, N50 = 242702, L50 = 8, Total length = 5443386, GC % = 57.38, # N's per 100 kbp =  0.00\n",
      "    24 SRR10512965, N50 = 211559, L50 = 8, Total length = 5568676, GC % = 57.17, # N's per 100 kbp =  0.00\n",
      "    25 SRR10512968, N50 = 274199, L50 = 7, Total length = 5580176, GC % = 57.17, # N's per 100 kbp =  0.00\n",
      "    26 SRR10513325, N50 = 15286, L50 = 53, Total length = 2823023, GC % = 38.14, # N's per 100 kbp =  0.00\n",
      "    27 SRR10513326, N50 = 475877, L50 = 3, Total length = 2918618, GC % = 37.83, # N's per 100 kbp =  0.00\n",
      "    28 SRR10513328, N50 = 24985, L50 = 36, Total length = 2812979, GC % = 38.16, # N's per 100 kbp =  0.00\n",
      "    29 SRR10513363, N50 = 30027, L50 = 29, Total length = 2911959, GC % = 38.05, # N's per 100 kbp =  0.00\n",
      "    30 SRR10513672, N50 = 408294, L50 = 2, Total length = 2897485, GC % = 37.88, # N's per 100 kbp =  0.00\n",
      "    31 SRR10519468, N50 = 43156, L50 = 13, Total length = 1675936, GC % = 30.52, # N's per 100 kbp =  0.00\n",
      "    32 SRR10519469, N50 = 43688, L50 = 10, Total length = 1634884, GC % = 31.45, # N's per 100 kbp =  0.00\n",
      "    33 SRR10519616, N50 = 183188, L50 = 9, Total length = 4681498, GC % = 52.13, # N's per 100 kbp =  0.00\n",
      "    34 SRR10519617, N50 = 78116, L50 = 19, Total length = 4485528, GC % = 52.42, # N's per 100 kbp =  0.00\n",
      "    35 SRR10519619, N50 = 180314, L50 = 9, Total length = 4885162, GC % = 52.08, # N's per 100 kbp =  0.00\n",
      "    36 SRR10519620, N50 = 116463, L50 = 14, Total length = 4673229, GC % = 52.18, # N's per 100 kbp =  0.00\n",
      "    37 SRR10519637, N50 = 196648, L50 = 10, Total length = 4685133, GC % = 52.13, # N's per 100 kbp =  0.00\n",
      "    38 SRR10521982, N50 = 51756, L50 = 12, Total length = 1682775, GC % = 30.30, # N's per 100 kbp =  0.00\n",
      "    39 SRR10521983, N50 = 82064, L50 = 8, Total length = 1785953, GC % = 30.10, # N's per 100 kbp =  0.00\n",
      "    40 SRR10521984, N50 = 104618, L50 = 6, Total length = 1749441, GC % = 31.18, # N's per 100 kbp =  0.00\n",
      "    41 SRR10527348, N50 = 23120, L50 = 59, Total length = 4311835, GC % = 50.94, # N's per 100 kbp =  0.00\n",
      "    42 SRR10527349, N50 = 21633, L50 = 64, Total length = 4475054, GC % = 50.68, # N's per 100 kbp =  0.00\n",
      "    43 SRR10527351, N50 = 22226, L50 = 62, Total length = 4631159, GC % = 50.64, # N's per 100 kbp =  0.00\n",
      "    44 SRR10527352, N50 = 49216, L50 = 28, Total length = 4833031, GC % = 50.63, # N's per 100 kbp =  0.00\n",
      "    45 SRR10527353, N50 = 51214, L50 = 26, Total length = 5086042, GC % = 50.54, # N's per 100 kbp =  0.00\n",
      "    46 SRR8088181, N50 = 40587, L50 = 16, Total length = 2141606, GC % = 52.44, # N's per 100 kbp =  0.00\n",
      "    47 SRR8088182, N50 = 39051, L50 = 18, Total length = 2141169, GC % = 52.41, # N's per 100 kbp =  0.00\n",
      "    48 SRR8088183, N50 = 40702, L50 = 17, Total length = 2064650, GC % = 52.64, # N's per 100 kbp =  0.00\n",
      "    49 SRR8088184, N50 = 35194, L50 = 21, Total length = 2106514, GC % = 52.61, # N's per 100 kbp =  0.00\n",
      "    50 SRR8088185, N50 = 40328, L50 = 16, Total length = 2107836, GC % = 52.58, # N's per 100 kbp =  0.00\n",
      "  Drawing Nx plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/Nx_plot.pdf\n",
      "  Drawing cumulative plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/cumulative_plot.pdf\n",
      "  Drawing GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/GC_content_plot.pdf\n",
      "  Drawing ERR1144974 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/ERR1144974_GC_content_plot.pdf\n",
      "  Drawing ERR1144975 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/ERR1144975_GC_content_plot.pdf\n",
      "  Drawing ERR1144976 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/ERR1144976_GC_content_plot.pdf\n",
      "  Drawing ERR1144977 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/ERR1144977_GC_content_plot.pdf\n",
      "  Drawing ERR1144978 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/ERR1144978_GC_content_plot.pdf\n",
      "  Drawing ERR3655992 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/ERR3655992_GC_content_plot.pdf\n",
      "  Drawing ERR3655994 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/ERR3655994_GC_content_plot.pdf\n",
      "  Drawing ERR3655996 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/ERR3655996_GC_content_plot.pdf\n",
      "  Drawing ERR3655998 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/ERR3655998_GC_content_plot.pdf\n",
      "  Drawing ERR3656002 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/ERR3656002_GC_content_plot.pdf\n",
      "  Drawing ERR3656004 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/ERR3656004_GC_content_plot.pdf\n",
      "  Drawing ERR3656010 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/ERR3656010_GC_content_plot.pdf\n",
      "  Drawing ERR3656012 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/ERR3656012_GC_content_plot.pdf\n",
      "  Drawing ERR3656013 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/ERR3656013_GC_content_plot.pdf\n",
      "  Drawing ERR3656015 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/ERR3656015_GC_content_plot.pdf\n",
      "  Drawing ERR3656018 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/ERR3656018_GC_content_plot.pdf\n",
      "  Drawing ERR3656019 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/ERR3656019_GC_content_plot.pdf\n",
      "  Drawing SRR10298903 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/SRR10298903_GC_content_plot.pdf\n",
      "  Drawing SRR10298904 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/SRR10298904_GC_content_plot.pdf\n",
      "  Drawing SRR10298905 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/SRR10298905_GC_content_plot.pdf\n",
      "  Drawing SRR10298906 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/SRR10298906_GC_content_plot.pdf\n",
      "  Drawing SRR10298907 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/SRR10298907_GC_content_plot.pdf\n",
      "  Drawing SRR10512964 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/SRR10512964_GC_content_plot.pdf\n",
      "  Drawing SRR10512965 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/SRR10512965_GC_content_plot.pdf\n",
      "  Drawing SRR10512968 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/SRR10512968_GC_content_plot.pdf\n",
      "  Drawing SRR10513325 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/SRR10513325_GC_content_plot.pdf\n",
      "  Drawing SRR10513326 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/SRR10513326_GC_content_plot.pdf\n",
      "  Drawing SRR10513328 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/SRR10513328_GC_content_plot.pdf\n",
      "  Drawing SRR10513363 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/SRR10513363_GC_content_plot.pdf\n",
      "  Drawing SRR10513672 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/SRR10513672_GC_content_plot.pdf\n",
      "  Drawing SRR10519468 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/SRR10519468_GC_content_plot.pdf\n",
      "  Drawing SRR10519469 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/SRR10519469_GC_content_plot.pdf\n",
      "  Drawing SRR10519616 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/SRR10519616_GC_content_plot.pdf\n",
      "  Drawing SRR10519617 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/SRR10519617_GC_content_plot.pdf\n",
      "  Drawing SRR10519619 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/SRR10519619_GC_content_plot.pdf\n",
      "  Drawing SRR10519620 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/SRR10519620_GC_content_plot.pdf\n",
      "  Drawing SRR10519637 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/SRR10519637_GC_content_plot.pdf\n",
      "  Drawing SRR10521982 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/SRR10521982_GC_content_plot.pdf\n",
      "  Drawing SRR10521983 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/SRR10521983_GC_content_plot.pdf\n",
      "  Drawing SRR10521984 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/SRR10521984_GC_content_plot.pdf\n",
      "  Drawing SRR10527348 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/SRR10527348_GC_content_plot.pdf\n",
      "  Drawing SRR10527349 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/SRR10527349_GC_content_plot.pdf\n",
      "  Drawing SRR10527351 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/SRR10527351_GC_content_plot.pdf\n",
      "  Drawing SRR10527352 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/SRR10527352_GC_content_plot.pdf\n",
      "  Drawing SRR10527353 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/SRR10527353_GC_content_plot.pdf\n",
      "  Drawing SRR8088181 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/SRR8088181_GC_content_plot.pdf\n",
      "  Drawing SRR8088182 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/SRR8088182_GC_content_plot.pdf\n",
      "  Drawing SRR8088183 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/SRR8088183_GC_content_plot.pdf\n",
      "  Drawing SRR8088184 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/SRR8088184_GC_content_plot.pdf\n",
      "  Drawing SRR8088185 GC content plot...\n",
      "    saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/basic_stats/SRR8088185_GC_content_plot.pdf\n",
      "Done.\n",
      "\n",
      "NOTICE: Genes are not predicted by default. Use --gene-finding or --glimmer option to enable it.\n",
      "\n",
      "2019-12-04 18:54:18\n",
      "Creating large visual summaries...\n",
      "This may take a while: press Ctrl-C to skip this step..\n",
      "  1 of 2: Creating Icarus viewers...\n",
      "  2 of 2: Creating PDF with all tables and plots...\n",
      "Done\n",
      "\n",
      "2019-12-04 18:54:26\n",
      "RESULTS:\n",
      "  Text versions of total report are saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/report.txt, report.tsv, and report.tex\n",
      "  Text versions of transposed total report are saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/transposed_report.txt, transposed_report.tsv, and transposed_report.tex\n",
      "  HTML version (interactive tables and plots) is saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/report.html\n",
      "  PDF version (tables and plots) is saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/report.pdf\n",
      "  Icarus (contig browser) is saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/icarus.html\n",
      "  Log is saved to /home/CSCScience.ca/apetkau/workspace/comp7934-project/microbial/assembly/quast/quast.log\n",
      "\n",
      "Finished: 2019-12-04 18:54:26\n",
      "Elapsed time: 0:00:27.268858\n",
      "NOTICEs: 2; WARNINGs: 0; non-fatal ERRORs: 0\n",
      "\n",
      "Thank you for using QUAST!\n"
     ]
    }
   ],
   "source": [
    "conda run --name quast quast -o assembly/quast assembly/*.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0m\u001b[01;34mbasic_stats\u001b[0m     quast.log    report.tex  transposed_report.tex\n",
      "icarus.html     report.html  report.tsv  transposed_report.tsv\n",
      "\u001b[01;34micarus_viewers\u001b[0m  report.pdf   report.txt  transposed_report.txt\n"
     ]
    }
   ],
   "source": [
    "ls assembly/quast"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 3: Search for AMR genes\n",
    "\n",
    "Let's use [staramr](https://github.com/phac-nml/staramr) to search for antimicrobial resistance genes in all assemblies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "staramr 0.7.0\n",
      "2019-12-04 18:47:33 INFO: No --pointfinder-organism specified. Will not search the PointFinder databases\n",
      "2019-12-04 18:47:33 INFO: No --plasmidfinder-database-type specified. Will search the entire PlasmidFinder database\n",
      "2019-12-04 18:47:33 INFO: --output-dir set. All files will be output to [assembly/staramr]\n",
      "2019-12-04 18:47:33 INFO: Will exclude ResFinder/PointFinder genes listed in [/home/CSCScience.ca/apetkau/miniconda3/envs/staramr/lib/python3.7/site-packages/staramr/databases/exclude/data/genes_to_exclude.tsv]. Use --no-exclude-genes to disable\n",
      "2019-12-04 18:47:33 INFO: Making BLAST databases for input files\n",
      "2019-12-04 18:47:34 INFO: Scheduling blasts and MLST for ERR1144974.fasta\n",
      "2019-12-04 18:47:34 INFO: Scheduling blasts and MLST for ERR1144975.fasta\n",
      "2019-12-04 18:47:34 INFO: Scheduling blasts and MLST for ERR1144976.fasta\n",
      "2019-12-04 18:47:34 INFO: Scheduling blasts and MLST for ERR1144977.fasta\n",
      "2019-12-04 18:47:34 INFO: Scheduling blasts and MLST for ERR1144978.fasta\n",
      "2019-12-04 18:47:35 INFO: Scheduling blasts and MLST for ERR3655992.fasta\n",
      "2019-12-04 18:47:35 INFO: Scheduling blasts and MLST for ERR3655994.fasta\n",
      "2019-12-04 18:47:35 INFO: Scheduling blasts and MLST for ERR3655996.fasta\n",
      "2019-12-04 18:47:35 INFO: Scheduling blasts and MLST for ERR3655998.fasta\n",
      "2019-12-04 18:47:35 INFO: Scheduling blasts and MLST for ERR3656002.fasta\n",
      "2019-12-04 18:47:35 INFO: Scheduling blasts and MLST for ERR3656004.fasta\n",
      "2019-12-04 18:47:35 INFO: Scheduling blasts and MLST for ERR3656010.fasta\n",
      "2019-12-04 18:47:36 INFO: Scheduling blasts and MLST for ERR3656012.fasta\n",
      "2019-12-04 18:47:36 INFO: Scheduling blasts and MLST for ERR3656013.fasta\n",
      "2019-12-04 18:47:36 INFO: Scheduling blasts and MLST for ERR3656015.fasta\n",
      "2019-12-04 18:47:36 INFO: Scheduling blasts and MLST for ERR3656018.fasta\n",
      "2019-12-04 18:47:36 INFO: Scheduling blasts and MLST for ERR3656019.fasta\n",
      "2019-12-04 18:47:36 INFO: Scheduling blasts and MLST for SRR10298903.fasta\n",
      "2019-12-04 18:47:36 INFO: Scheduling blasts and MLST for SRR10298904.fasta\n",
      "2019-12-04 18:47:37 INFO: Scheduling blasts and MLST for SRR10298905.fasta\n",
      "2019-12-04 18:47:37 INFO: Scheduling blasts and MLST for SRR10298906.fasta\n",
      "2019-12-04 18:47:37 INFO: Scheduling blasts and MLST for SRR10298907.fasta\n",
      "2019-12-04 18:47:37 INFO: Scheduling blasts and MLST for SRR10512964.fasta\n",
      "2019-12-04 18:47:37 INFO: Scheduling blasts and MLST for SRR10512965.fasta\n",
      "2019-12-04 18:47:37 INFO: Scheduling blasts and MLST for SRR10512968.fasta\n",
      "2019-12-04 18:47:37 INFO: Scheduling blasts and MLST for SRR10513325.fasta\n",
      "2019-12-04 18:47:38 INFO: Scheduling blasts and MLST for SRR10513326.fasta\n",
      "2019-12-04 18:47:38 INFO: Scheduling blasts and MLST for SRR10513328.fasta\n",
      "2019-12-04 18:47:38 INFO: Scheduling blasts and MLST for SRR10513363.fasta\n",
      "2019-12-04 18:47:38 INFO: Scheduling blasts and MLST for SRR10513672.fasta\n",
      "2019-12-04 18:47:38 INFO: Scheduling blasts and MLST for SRR10519468.fasta\n",
      "2019-12-04 18:47:38 INFO: Scheduling blasts and MLST for SRR10519469.fasta\n",
      "2019-12-04 18:47:39 INFO: Scheduling blasts and MLST for SRR10519616.fasta\n",
      "2019-12-04 18:47:39 INFO: Scheduling blasts and MLST for SRR10519617.fasta\n",
      "2019-12-04 18:47:39 INFO: Scheduling blasts and MLST for SRR10519619.fasta\n",
      "2019-12-04 18:47:39 INFO: Scheduling blasts and MLST for SRR10519620.fasta\n",
      "2019-12-04 18:47:39 INFO: Scheduling blasts and MLST for SRR10519637.fasta\n",
      "2019-12-04 18:47:39 INFO: Scheduling blasts and MLST for SRR10521982.fasta\n",
      "2019-12-04 18:47:39 INFO: Scheduling blasts and MLST for SRR10521983.fasta\n",
      "2019-12-04 18:47:40 INFO: Scheduling blasts and MLST for SRR10521984.fasta\n",
      "2019-12-04 18:47:40 INFO: Scheduling blasts and MLST for SRR10527348.fasta\n",
      "2019-12-04 18:47:40 INFO: Scheduling blasts and MLST for SRR10527349.fasta\n",
      "2019-12-04 18:47:40 INFO: Scheduling blasts and MLST for SRR10527351.fasta\n",
      "2019-12-04 18:47:40 INFO: Scheduling blasts and MLST for SRR10527352.fasta\n",
      "2019-12-04 18:47:40 INFO: Scheduling blasts and MLST for SRR10527353.fasta\n",
      "2019-12-04 18:47:40 INFO: Scheduling blasts and MLST for SRR8088181.fasta\n",
      "2019-12-04 18:47:41 INFO: Scheduling blasts and MLST for SRR8088182.fasta\n",
      "2019-12-04 18:47:41 INFO: Scheduling blasts and MLST for SRR8088183.fasta\n",
      "2019-12-04 18:47:41 INFO: Scheduling blasts and MLST for SRR8088184.fasta\n",
      "2019-12-04 18:47:41 INFO: Scheduling blasts and MLST for SRR8088185.fasta\n",
      "2019-12-04 18:47:43 WARNING: No drug found for drug_class=macrolide, gene=erm(37)_1, accession=AL123456\n",
      "2019-12-04 18:47:43 WARNING: No drug found for drug_class=macrolide, gene=erm(37)_1, accession=AL123456\n",
      "2019-12-04 18:47:43 WARNING: No drug found for drug_class=macrolide, gene=erm(37)_1, accession=AL123456\n",
      "2019-12-04 18:47:43 WARNING: No drug found for drug_class=macrolide, gene=erm(37)_1, accession=AL123456\n",
      "2019-12-04 18:47:43 WARNING: No drug found for drug_class=macrolide, gene=erm(37)_1, accession=AL123456\n",
      "2019-12-04 18:47:43 WARNING: No drug found for drug_class=phenicol, gene=catB4_1, accession=AF322577\n",
      "2019-12-04 18:47:43 WARNING: No drug found for drug_class=quinolone, gene=crpP_1, accession=HM560971\n",
      "2019-12-04 18:47:43 WARNING: No drug found for drug_class=quinolone, gene=crpP_1, accession=HM560971\n",
      "2019-12-04 18:47:43 WARNING: No drug found for drug_class=phenicol, gene=catB4_1, accession=AF322577\n",
      "2019-12-04 18:47:43 WARNING: No drug found for drug_class=quinolone, gene=crpP_1, accession=HM560971\n",
      "2019-12-04 18:47:44 WARNING: No drug found for drug_class=phenicol, gene=catB4_1, accession=AF322577\n",
      "2019-12-04 18:47:45 WARNING: No drug found for drug_class=phenicol, gene=catB4_1, accession=AF322577\n",
      "2019-12-04 18:47:46 WARNING: No drug found for drug_class=phenicol, gene=catB4_1, accession=AF322577\n",
      "2019-12-04 18:47:51 INFO: Finished. Took 0.31 minutes.\n",
      "2019-12-04 18:47:51 INFO: Predicting AMR resistance phenotypes is enabled. The predictions are for microbiological resistance and *not* clinical resistance. These results are continually being improved and we welcome any feedback.\n",
      "2019-12-04 18:47:51 INFO: Writing resfinder to [assembly/staramr/resfinder.tsv]\n",
      "2019-12-04 18:47:51 INFO: --output-dir or --output-pointfinder unset. No pointfinder file will be written\n",
      "2019-12-04 18:47:51 INFO: Writing plasmidfinder to [assembly/staramr/plasmidfinder.tsv]\n",
      "2019-12-04 18:47:51 INFO: Writing summary to [assembly/staramr/summary.tsv]\n",
      "2019-12-04 18:47:51 INFO: Writing MLST summary to [assembly/staramr/mlst.tsv]\n",
      "2019-12-04 18:47:51 INFO: Writing detailed summary to [assembly/staramr/detailed_summary.tsv]\n",
      "2019-12-04 18:47:51 INFO: Writing settings to [assembly/staramr/settings.txt]\n",
      "2019-12-04 18:47:51 INFO: Writing Excel to [assembly/staramr/results.xlsx]\n",
      "2019-12-04 18:47:52 INFO: BLAST hits are stored in [assembly/staramr/hits]\n"
     ]
    }
   ],
   "source": [
    "conda run --name staramr staramr --version\n",
    "conda run --name staramr staramr search -o assembly/staramr --exclude-negatives assembly/*.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's take a look at the results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Isolate ID   Plasmid           %Identity  %Overlap  HSP Length/Total Length  Contig                 Start  End    Accession\n",
      "ERR3655994   Col440II          100.00     100.00    282/282                  Contig_2_23.4654_Circ  3395   3676   CP023921.1\n",
      "ERR3655996   IncQ1             100.00     64.20     511/796                  Contig_71_5.84164      511    1      M28829.1\n",
      "ERR3655998   Col440II          100.00     100.00    282/282                  Contig_5_18.1039       4555   4836   CP023921.1\n",
      "ERR3656002   IncI              100.00     97.16     137/141                  Contig_69_7.05375      235    371    AP011954\n",
      "ERR3656018   ColRNAI           99.23      100.00    130/130                  Contig_3_159.311_Circ  1104   975    DQ298019\n",
      "ERR3656018   IncFIA(HI1)       98.45      99.74     387/388                  Contig_40_61.9395      5334   4948   AF250878\n",
      "ERR3656018   IncFIB(K)         100.00     100.00    560/560                  Contig_50_53.1909      8451   7892   JN233704\n",
      "ERR3656018   IncFII(K)         100.00     100.00    148/148                  Contig_117_49.5379     59563  59710  CP000648\n",
      "ERR3656018   IncN              99.81      100.00    514/514                  Contig_7_61.8497       744    231    AY046276\n",
      "ERR3656018   IncR              100.00     100.00    251/251                  Contig_32_58.2535      5398   5648   DQ449578\n",
      "ERR3656018   rep11             100.00     100.00    945/945                  Contig_39_17.4598      2180   3124   AB178871\n",
      "ERR3656018   rep18             99.43      93.44     527/564                  Contig_1_14.497_Circ   2399   2925   AF408195\n",
      "ERR3656018   rep22             100.00     63.84     452/708                  Contig_73_7.93145      550    1001   X03408.1\n",
      "SRR10512964  ColRNAI           100.00     100.00    130/130                  Contig_1_103.881_Circ  6572   6701   DQ298019\n",
      "SRR10512964  IncR              100.00     100.00    251/251                  Contig_64_57.6595      5358   5608   DQ449578\n",
      "SRR10512965  IncN              99.81      100.00    514/514                  Contig_50_128.483      1018   505    AY046276\n",
      "SRR10512965  IncR              99.20      100.00    251/251                  Contig_73_33.3514      11080  10830  DQ449578\n",
      "SRR10512968  IncN              99.81      100.00    514/514                  Contig_32_182.949      1035   522    AY046276\n",
      "SRR10512968  IncR              99.20      100.00    251/251                  Contig_30_44.2487      9791   9541   DQ449578\n",
      "SRR10519616  IncFIB(S)         98.91      100.00    643/643                  Contig_22_52.7116      48067  48709  FN432031\n",
      "SRR10519616  IncFII(S)         100.00     100.00    262/262                  Contig_22_52.7116      4075   3814   CP000858\n",
      "SRR10519619  IncFIB(S)         98.91      100.00    643/643                  Contig_37_48.3437      49285  49927  FN432031\n",
      "SRR10519619  IncFII(S)         100.00     100.00    262/262                  Contig_37_48.3437      5293   5032   CP000858\n",
      "SRR10519620  IncFIB(S)         98.91      100.00    643/643                  Contig_54_45.1151      48668  49310  FN432031\n",
      "SRR10519620  IncFII(S)         100.00     100.00    262/262                  Contig_54_45.1151      4677   4416   CP000858\n",
      "SRR10519637  IncFIB(S)         98.91      100.00    643/643                  Contig_34_52.8597      48669  49311  FN432031\n",
      "SRR10519637  IncFII(S)         100.00     100.00    262/262                  Contig_34_52.8597      4677   4416   CP000858\n",
      "SRR10527348  Col(BS512)        100.00     100.00    233/233                  Contig_3_334.758_Circ  632    400    010656\n",
      "SRR10527348  Col156            98.68      98.70     152/154                  Contig_205_97.0705     152    1      009781\n",
      "SRR10527349  IncFII            100.00     100.00    261/261                  Contig_54_50.3639      3322   3582   AY458016\n",
      "SRR10527351  Col(BS512)        100.00     100.00    233/233                  Contig_4_180.979_Circ  632    400    010656\n",
      "SRR10527351  Col156            99.19      79.87     123/154                  Contig_16_59.9034      281    403    009781\n",
      "SRR10527351  IncFII            100.00     100.00    261/261                  Contig_70_55.2737      3322   3582   AY458016\n",
      "SRR10527351  IncI1             100.00     100.00    142/142                  Contig_439_68.0809     7661   7520   AP005147\n",
      "SRR10527352  Col(BS512)        100.00     100.00    233/233                  Contig_2_325.425_Circ  632    400    010656\n",
      "SRR10527353  IncFIB(AP001918)  98.68      100.00    682/682                  Contig_210_57.1196     865    1546   AP001918\n",
      "SRR10527353  IncFII(pRSB107)   100.00     100.00    261/261                  Contig_263_53.8092     9133   8873   AJ851089\n"
     ]
    }
   ],
   "source": [
    "column -s$'\\t' -t assembly/staramr/plasmidfinder.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This gives us a list of all plasmids (specifically plasmid incompatibility factors) that were found using BLAST. We can make use of these to test out BIGSI/HowDeSBT.\n",
    "\n",
    "For example, let's take `Col440II`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Isolate ID  Plasmid   %Identity  %Overlap  HSP Length/Total Length  Contig                 Start  End   Accession\n",
      "ERR3655994  Col440II  100.00     100.00    282/282                  Contig_2_23.4654_Circ  3395   3676  CP023921.1\n",
      "ERR3655998  Col440II  100.00     100.00    282/282                  Contig_5_18.1039       4555   4836  CP023921.1\n"
     ]
    }
   ],
   "source": [
    "(head -n 1 assembly/staramr/plasmidfinder.tsv; grep 'Col440II' assembly/staramr/plasmidfinder.tsv) | column -s$'\\t' -t"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There's only two samples where this is found. So, if we take the specfic sequence found:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">\u001b[01;31m\u001b[KCol440II\u001b[m\u001b[K_1__CP023921.1 isolate: ERR3655994, contig: Contig_2_23.4654_Circ, contig_start: 3395, contig_end: 3676, database_gene_start: 1, database_gene_end: 282, hsp/length: 282/282, pid: 100.00%, plength: 100.00%\n",
      "GTCGATTGCCATCAGTGCGGCCACAATCTGCACCCGGTCATGACCGGCACCACCGGCATT\n",
      "CACTTTCCGGGCGATCTGGTTCAGGTTGTTCCCCATACCGGCAAGCTGGCGCAGCAGCGC\n",
      "CGGCGATATCGACGGCAGTCTGCCGGCACGCGCCGGCTTCTCATCGAGGCAGGTCTGACG\n",
      "CATCCACGCCGCCAGCTGCTTGCCGTCGCACCGCTCGAGCAGCCGCCGGTGTTCGTCTTC\n",
      "CGTCACCCACATCGTGAGCATCTTGTTGCGTTTGTCTGCCAG\n"
     ]
    }
   ],
   "source": [
    "grep -A5 'Col440II' assembly/staramr/hits/plasmidfinder_ERR3655994.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "we can use this sequence in the BIGSI/HowDeSBT searches to test if they only find it in the two above genomes."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
