{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0m\u001b[01;34mdata\u001b[0m  microbial-genomes.txt  README.md  Untitled.ipynb\n"
     ]
    }
   ],
   "source": [
    "PROJECT_DIR=`git rev-parse --show-toplevel`\n",
    "cd $PROJECT_DIR/microbial\n",
    "ls"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 1: Download data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘data’: File exists\n",
      "+ fasterq-dump -o data/SRR10519619 --split-files SRR10519619\n",
      "spots read      : 2,465,304\n",
      "reads read      : 4,930,608\n",
      "reads written   : 4,930,608\n",
      "+ pigz data/SRR10519619_1.fastq data/SRR10519619_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/SRR10519637 --split-files SRR10519637\n",
      "spots read      : 7,316,985\n",
      "reads read      : 14,633,970\n",
      "reads written   : 14,633,970\n",
      "+ pigz data/SRR10519637_1.fastq data/SRR10519637_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/SRR10519620 --split-files SRR10519620\n",
      "spots read      : 3,443,404\n",
      "reads read      : 6,886,808\n",
      "reads written   : 6,886,808\n",
      "+ pigz data/SRR10519620_1.fastq data/SRR10519620_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/SRR10519617 --split-files SRR10519617\n",
      "spots read      : 2,681,834\n",
      "reads read      : 5,363,668\n",
      "reads written   : 5,363,668\n",
      "+ pigz data/SRR10519617_1.fastq data/SRR10519617_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/SRR10519616 --split-files SRR10519616\n",
      "spots read      : 2,998,025\n",
      "reads read      : 5,996,050\n",
      "reads written   : 5,996,050\n",
      "+ pigz data/SRR10519616_1.fastq data/SRR10519616_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/SRR10527352 --split-files SRR10527352\n",
      "spots read      : 1,638,033\n",
      "reads read      : 3,276,066\n",
      "reads written   : 3,276,066\n",
      "+ pigz data/SRR10527352_1.fastq data/SRR10527352_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/SRR10527353 --split-files SRR10527353\n",
      "spots read      : 1,832,803\n",
      "reads read      : 3,665,606\n",
      "reads written   : 3,665,606\n",
      "+ pigz data/SRR10527353_1.fastq data/SRR10527353_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/SRR10527348 --split-files SRR10527348\n",
      "spots read      : 2,281,618\n",
      "reads read      : 4,563,236\n",
      "reads written   : 4,563,236\n",
      "+ pigz data/SRR10527348_1.fastq data/SRR10527348_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/SRR10527351 --split-files SRR10527351\n",
      "spots read      : 1,764,396\n",
      "reads read      : 3,528,792\n",
      "reads written   : 3,528,792\n",
      "+ pigz data/SRR10527351_1.fastq data/SRR10527351_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/SRR10527349 --split-files SRR10527349\n",
      "spots read      : 1,649,218\n",
      "reads read      : 3,298,436\n",
      "reads written   : 3,298,436\n",
      "+ pigz data/SRR10527349_1.fastq data/SRR10527349_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/SRR10521984 --split-files SRR10521984\n",
      "spots read      : 1,823,660\n",
      "reads read      : 3,647,320\n",
      "reads written   : 3,647,320\n",
      "+ pigz data/SRR10521984_1.fastq data/SRR10521984_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/SRR10521982 --split-files SRR10521982\n",
      "spots read      : 2,542,008\n",
      "reads read      : 5,084,016\n",
      "reads written   : 5,084,016\n",
      "+ pigz data/SRR10521982_1.fastq data/SRR10521982_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/SRR10521983 --split-files SRR10521983\n",
      "spots read      : 2,160,344\n",
      "reads read      : 4,320,688\n",
      "reads written   : 4,320,688\n",
      "+ pigz data/SRR10521983_1.fastq data/SRR10521983_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/SRR10519469 --split-files SRR10519469\n",
      "spots read      : 3,124,819\n",
      "reads read      : 6,249,638\n",
      "reads written   : 6,249,638\n",
      "+ pigz data/SRR10519469_1.fastq data/SRR10519469_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/SRR10519468 --split-files SRR10519468\n",
      "spots read      : 4,164,792\n",
      "reads read      : 8,329,584\n",
      "reads written   : 8,329,584\n",
      "+ pigz data/SRR10519468_1.fastq data/SRR10519468_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/SRR10513672 --split-files SRR10513672\n",
      "spots read      : 903,007\n",
      "reads read      : 1,806,014\n",
      "reads written   : 1,806,014\n",
      "+ pigz data/SRR10513672_1.fastq data/SRR10513672_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/SRR10513363 --split-files SRR10513363\n",
      "spots read      : 4,180,524\n",
      "reads read      : 8,361,048\n",
      "reads written   : 8,361,048\n",
      "+ pigz data/SRR10513363_1.fastq data/SRR10513363_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/SRR10513326 --split-files SRR10513326\n",
      "spots read      : 1,838,335\n",
      "reads read      : 3,676,670\n",
      "reads written   : 3,676,670\n",
      "+ pigz data/SRR10513326_1.fastq data/SRR10513326_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/SRR10513325 --split-files SRR10513325\n",
      "spots read      : 1,203,226\n",
      "reads read      : 2,406,452\n",
      "reads written   : 2,406,452\n",
      "+ pigz data/SRR10513325_1.fastq data/SRR10513325_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/SRR10513328 --split-files SRR10513328\n",
      "spots read      : 1,882,971\n",
      "reads read      : 3,765,942\n",
      "reads written   : 3,765,942\n",
      "+ pigz data/SRR10513328_1.fastq data/SRR10513328_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/SRR10512968 --split-files SRR10512968\n",
      "spots read      : 1,123,096\n",
      "reads read      : 2,246,192\n",
      "reads written   : 2,246,192\n",
      "+ pigz data/SRR10512968_1.fastq data/SRR10512968_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/SRR10512964 --split-files SRR10512964\n",
      "spots read      : 999,107\n",
      "reads read      : 1,998,214\n",
      "reads written   : 1,998,214\n",
      "+ pigz data/SRR10512964_1.fastq data/SRR10512964_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/SRR10512965 --split-files SRR10512965\n",
      "spots read      : 543,952\n",
      "reads read      : 1,087,904\n",
      "reads written   : 1,087,904\n",
      "+ pigz data/SRR10512965_1.fastq data/SRR10512965_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/ERR3656018 --split-files ERR3656018\n",
      "spots read      : 2,512,546\n",
      "reads read      : 5,025,092\n",
      "reads written   : 5,025,092\n",
      "+ pigz data/ERR3656018_1.fastq data/ERR3656018_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/ERR3656019 --split-files ERR3656019\n",
      "spots read      : 2,458,314\n",
      "reads read      : 4,916,628\n",
      "reads written   : 4,916,628\n",
      "+ pigz data/ERR3656019_1.fastq data/ERR3656019_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/ERR1144974 --split-files ERR1144974\n",
      "spots read      : 18,937,653\n",
      "reads read      : 37,875,306\n",
      "reads written   : 37,875,306\n",
      "+ pigz data/ERR1144974_1.fastq data/ERR1144974_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/ERR1144975 --split-files ERR1144975\n",
      "spots read      : 11,701,339\n",
      "reads read      : 23,402,678\n",
      "reads written   : 23,402,678\n",
      "+ pigz data/ERR1144975_1.fastq data/ERR1144975_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/ERR1144976 --split-files ERR1144976\n",
      "spots read      : 9,687,071\n",
      "reads read      : 19,374,142\n",
      "reads written   : 19,374,142\n",
      "+ pigz data/ERR1144976_1.fastq data/ERR1144976_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/ERR1144977 --split-files ERR1144977\n",
      "spots read      : 12,010,912\n",
      "reads read      : 24,021,824\n",
      "reads written   : 24,021,824\n",
      "+ pigz data/ERR1144977_1.fastq data/ERR1144977_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/ERR1144978 --split-files ERR1144978\n",
      "spots read      : 11,161,674\n",
      "reads read      : 22,323,348\n",
      "reads written   : 22,323,348\n",
      "+ pigz data/ERR1144978_1.fastq data/ERR1144978_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/SRR8088185 --split-files SRR8088185\n",
      "spots read      : 762,692\n",
      "reads read      : 1,525,384\n",
      "reads written   : 1,525,384\n",
      "+ pigz data/SRR8088185_1.fastq data/SRR8088185_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/SRR8088181 --split-files SRR8088181\n",
      "spots read      : 715,282\n",
      "reads read      : 1,430,564\n",
      "reads written   : 1,430,564\n",
      "+ pigz data/SRR8088181_1.fastq data/SRR8088181_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/SRR8088184 --split-files SRR8088184\n",
      "spots read      : 530,412\n",
      "reads read      : 1,060,824\n",
      "reads written   : 1,060,824\n",
      "+ pigz data/SRR8088184_1.fastq data/SRR8088184_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/SRR8088182 --split-files SRR8088182\n",
      "spots read      : 586,288\n",
      "reads read      : 1,172,576\n",
      "reads written   : 1,172,576\n",
      "+ pigz data/SRR8088182_1.fastq data/SRR8088182_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/SRR8088183 --split-files SRR8088183\n",
      "spots read      : 615,101\n",
      "reads read      : 1,230,202\n",
      "reads written   : 1,230,202\n",
      "+ pigz data/SRR8088183_1.fastq data/SRR8088183_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/ERR3656013 --split-files ERR3656013\n",
      "spots read      : 2,204,984\n",
      "reads read      : 4,409,968\n",
      "reads written   : 4,409,968\n",
      "+ pigz data/ERR3656013_1.fastq data/ERR3656013_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/ERR3656015 --split-files ERR3656015\n",
      "spots read      : 1,876,686\n",
      "reads read      : 3,753,372\n",
      "reads written   : 3,753,372\n",
      "+ pigz data/ERR3656015_1.fastq data/ERR3656015_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/ERR3656010 --split-files ERR3656010\n",
      "spots read      : 2,572,532\n",
      "reads read      : 5,145,064\n",
      "reads written   : 5,145,064\n",
      "+ pigz data/ERR3656010_1.fastq data/ERR3656010_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/ERR3656012 --split-files ERR3656012\n",
      "spots read      : 1,949,899\n",
      "reads read      : 3,899,798\n",
      "reads written   : 3,899,798\n",
      "+ pigz data/ERR3656012_1.fastq data/ERR3656012_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/ERR3656004 --split-files ERR3656004\n",
      "spots read      : 2,285,720\n",
      "reads read      : 4,571,440\n",
      "reads written   : 4,571,440\n",
      "+ pigz data/ERR3656004_1.fastq data/ERR3656004_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/ERR3656002 --split-files ERR3656002\n",
      "spots read      : 3,135,157\n",
      "reads read      : 6,270,314\n",
      "reads written   : 6,270,314\n",
      "+ pigz data/ERR3656002_1.fastq data/ERR3656002_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/ERR3655998 --split-files ERR3655998\n",
      "spots read      : 2,861,433\n",
      "reads read      : 5,722,866\n",
      "reads written   : 5,722,866\n",
      "+ pigz data/ERR3655998_1.fastq data/ERR3655998_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/ERR3655996 --split-files ERR3655996\n",
      "spots read      : 2,794,004\n",
      "reads read      : 5,588,008\n",
      "reads written   : 5,588,008\n",
      "+ pigz data/ERR3655996_1.fastq data/ERR3655996_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/ERR3655994 --split-files ERR3655994\n",
      "spots read      : 2,550,255\n",
      "reads read      : 5,100,510\n",
      "reads written   : 5,100,510\n",
      "+ pigz data/ERR3655994_1.fastq data/ERR3655994_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/ERR3655992 --split-files ERR3655992\n",
      "spots read      : 2,470,001\n",
      "reads read      : 4,940,002\n",
      "reads written   : 4,940,002\n",
      "+ pigz data/ERR3655992_1.fastq data/ERR3655992_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/SRR10298905 --split-files SRR10298905\n",
      "spots read      : 631,108\n",
      "reads read      : 1,262,216\n",
      "reads written   : 1,262,216\n",
      "+ pigz data/SRR10298905_1.fastq data/SRR10298905_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/SRR10298906 --split-files SRR10298906\n",
      "spots read      : 586,439\n",
      "reads read      : 1,172,878\n",
      "reads written   : 1,172,878\n",
      "+ pigz data/SRR10298906_1.fastq data/SRR10298906_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/SRR10298907 --split-files SRR10298907\n",
      "spots read      : 693,588\n",
      "reads read      : 1,387,176\n",
      "reads written   : 1,387,176\n",
      "+ pigz data/SRR10298907_1.fastq data/SRR10298907_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/SRR10298903 --split-files SRR10298903\n",
      "spots read      : 447,914\n",
      "reads read      : 895,828\n",
      "reads written   : 895,828\n",
      "+ pigz data/SRR10298903_1.fastq data/SRR10298903_2.fastq\n",
      "+ for name in `cat microbial-genomes.txt`\n",
      "+ set -x\n",
      "+ fasterq-dump -o data/SRR10298904 --split-files SRR10298904\n",
      "spots read      : 311,339\n",
      "reads read      : 622,678\n",
      "reads written   : 622,678\n",
      "+ pigz data/SRR10298904_1.fastq data/SRR10298904_2.fastq\n",
      "+ set +x\n"
     ]
    }
   ],
   "source": [
    "mkdir data\n",
    "\n",
    "for name in `cat microbial-genomes.txt`;\n",
    "do\n",
    "    set -x\n",
    "    fasterq-dump -o data/${name} --split-files ${name}\n",
    "    pigz data/${name}*\n",
    "done\n",
    "set +x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0m\u001b[01;31mERR1144974_1.fastq.gz\u001b[0m  \u001b[01;31mSRR10298903_1.fastq.gz\u001b[0m  \u001b[01;31mSRR10519619_1.fastq.gz\u001b[0m\n",
      "\u001b[01;31mERR1144974_2.fastq.gz\u001b[0m  \u001b[01;31mSRR10298903_2.fastq.gz\u001b[0m  \u001b[01;31mSRR10519619_2.fastq.gz\u001b[0m\n",
      "\u001b[01;31mERR1144975_1.fastq.gz\u001b[0m  \u001b[01;31mSRR10298904_1.fastq.gz\u001b[0m  \u001b[01;31mSRR10519620_1.fastq.gz\u001b[0m\n",
      "\u001b[01;31mERR1144975_2.fastq.gz\u001b[0m  \u001b[01;31mSRR10298904_2.fastq.gz\u001b[0m  \u001b[01;31mSRR10519620_2.fastq.gz\u001b[0m\n",
      "\u001b[01;31mERR1144976_1.fastq.gz\u001b[0m  \u001b[01;31mSRR10298905_1.fastq.gz\u001b[0m  \u001b[01;31mSRR10519637_1.fastq.gz\u001b[0m\n",
      "\u001b[01;31mERR1144976_2.fastq.gz\u001b[0m  \u001b[01;31mSRR10298905_2.fastq.gz\u001b[0m  \u001b[01;31mSRR10519637_2.fastq.gz\u001b[0m\n",
      "\u001b[01;31mERR1144977_1.fastq.gz\u001b[0m  \u001b[01;31mSRR10298906_1.fastq.gz\u001b[0m  \u001b[01;31mSRR10521982_1.fastq.gz\u001b[0m\n",
      "\u001b[01;31mERR1144977_2.fastq.gz\u001b[0m  \u001b[01;31mSRR10298906_2.fastq.gz\u001b[0m  \u001b[01;31mSRR10521982_2.fastq.gz\u001b[0m\n",
      "\u001b[01;31mERR1144978_1.fastq.gz\u001b[0m  \u001b[01;31mSRR10298907_1.fastq.gz\u001b[0m  \u001b[01;31mSRR10521983_1.fastq.gz\u001b[0m\n",
      "\u001b[01;31mERR1144978_2.fastq.gz\u001b[0m  \u001b[01;31mSRR10298907_2.fastq.gz\u001b[0m  \u001b[01;31mSRR10521983_2.fastq.gz\u001b[0m\n",
      "\u001b[01;31mERR3655992_1.fastq.gz\u001b[0m  \u001b[01;31mSRR10512964_1.fastq.gz\u001b[0m  \u001b[01;31mSRR10521984_1.fastq.gz\u001b[0m\n",
      "\u001b[01;31mERR3655992_2.fastq.gz\u001b[0m  \u001b[01;31mSRR10512964_2.fastq.gz\u001b[0m  \u001b[01;31mSRR10521984_2.fastq.gz\u001b[0m\n",
      "\u001b[01;31mERR3655994_1.fastq.gz\u001b[0m  \u001b[01;31mSRR10512965_1.fastq.gz\u001b[0m  \u001b[01;31mSRR10527348_1.fastq.gz\u001b[0m\n",
      "\u001b[01;31mERR3655994_2.fastq.gz\u001b[0m  \u001b[01;31mSRR10512965_2.fastq.gz\u001b[0m  \u001b[01;31mSRR10527348_2.fastq.gz\u001b[0m\n",
      "\u001b[01;31mERR3655996_1.fastq.gz\u001b[0m  \u001b[01;31mSRR10512968_1.fastq.gz\u001b[0m  \u001b[01;31mSRR10527349_1.fastq.gz\u001b[0m\n",
      "\u001b[01;31mERR3655996_2.fastq.gz\u001b[0m  \u001b[01;31mSRR10512968_2.fastq.gz\u001b[0m  \u001b[01;31mSRR10527349_2.fastq.gz\u001b[0m\n",
      "\u001b[01;31mERR3655998_1.fastq.gz\u001b[0m  \u001b[01;31mSRR10513325_1.fastq.gz\u001b[0m  \u001b[01;31mSRR10527351_1.fastq.gz\u001b[0m\n",
      "\u001b[01;31mERR3655998_2.fastq.gz\u001b[0m  \u001b[01;31mSRR10513325_2.fastq.gz\u001b[0m  \u001b[01;31mSRR10527351_2.fastq.gz\u001b[0m\n",
      "\u001b[01;31mERR3656002_1.fastq.gz\u001b[0m  \u001b[01;31mSRR10513326_1.fastq.gz\u001b[0m  \u001b[01;31mSRR10527352_1.fastq.gz\u001b[0m\n",
      "\u001b[01;31mERR3656002_2.fastq.gz\u001b[0m  \u001b[01;31mSRR10513326_2.fastq.gz\u001b[0m  \u001b[01;31mSRR10527352_2.fastq.gz\u001b[0m\n",
      "\u001b[01;31mERR3656004_1.fastq.gz\u001b[0m  \u001b[01;31mSRR10513328_1.fastq.gz\u001b[0m  \u001b[01;31mSRR10527353_1.fastq.gz\u001b[0m\n",
      "\u001b[01;31mERR3656004_2.fastq.gz\u001b[0m  \u001b[01;31mSRR10513328_2.fastq.gz\u001b[0m  \u001b[01;31mSRR10527353_2.fastq.gz\u001b[0m\n",
      "\u001b[01;31mERR3656010_1.fastq.gz\u001b[0m  \u001b[01;31mSRR10513363_1.fastq.gz\u001b[0m  \u001b[01;31mSRR8088181_1.fastq.gz\u001b[0m\n",
      "\u001b[01;31mERR3656010_2.fastq.gz\u001b[0m  \u001b[01;31mSRR10513363_2.fastq.gz\u001b[0m  \u001b[01;31mSRR8088181_2.fastq.gz\u001b[0m\n",
      "\u001b[01;31mERR3656012_1.fastq.gz\u001b[0m  \u001b[01;31mSRR10513672_1.fastq.gz\u001b[0m  \u001b[01;31mSRR8088182_1.fastq.gz\u001b[0m\n",
      "\u001b[01;31mERR3656012_2.fastq.gz\u001b[0m  \u001b[01;31mSRR10513672_2.fastq.gz\u001b[0m  \u001b[01;31mSRR8088182_2.fastq.gz\u001b[0m\n",
      "\u001b[01;31mERR3656013_1.fastq.gz\u001b[0m  \u001b[01;31mSRR10519468_1.fastq.gz\u001b[0m  \u001b[01;31mSRR8088183_1.fastq.gz\u001b[0m\n",
      "\u001b[01;31mERR3656013_2.fastq.gz\u001b[0m  \u001b[01;31mSRR10519468_2.fastq.gz\u001b[0m  \u001b[01;31mSRR8088183_2.fastq.gz\u001b[0m\n",
      "\u001b[01;31mERR3656015_1.fastq.gz\u001b[0m  \u001b[01;31mSRR10519469_1.fastq.gz\u001b[0m  \u001b[01;31mSRR8088184_1.fastq.gz\u001b[0m\n",
      "\u001b[01;31mERR3656015_2.fastq.gz\u001b[0m  \u001b[01;31mSRR10519469_2.fastq.gz\u001b[0m  \u001b[01;31mSRR8088184_2.fastq.gz\u001b[0m\n",
      "\u001b[01;31mERR3656018_1.fastq.gz\u001b[0m  \u001b[01;31mSRR10519616_1.fastq.gz\u001b[0m  \u001b[01;31mSRR8088185_1.fastq.gz\u001b[0m\n",
      "\u001b[01;31mERR3656018_2.fastq.gz\u001b[0m  \u001b[01;31mSRR10519616_2.fastq.gz\u001b[0m  \u001b[01;31mSRR8088185_2.fastq.gz\u001b[0m\n",
      "\u001b[01;31mERR3656019_1.fastq.gz\u001b[0m  \u001b[01;31mSRR10519617_1.fastq.gz\u001b[0m\n",
      "\u001b[01;31mERR3656019_2.fastq.gz\u001b[0m  \u001b[01;31mSRR10519617_2.fastq.gz\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "ls data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20G\tdata\n"
     ]
    }
   ],
   "source": [
    "du -sh data"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash [conda env:comp7934-project] *",
   "language": "bash",
   "name": "conda-env-comp7934-project-bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
